### ISSUE TITLE
Request: Enforce space after colon (`:`) for special comments

### ISSUE THREAD

**Issue by Avasam** [astral-sh/ruff#10160]
I've open the same request over at https://github.com/PyCQA/flake8-pyi/issues/465 , but I think that Ruff could generalize this, or at least expand on supported keywords. 

`# pyright:ignore` --> `# pyright: ignore`
`# type:ignore` --> `# type: ignore`
`# noqa:Y011` --> `# noqa: Y011`
`# fmt:on` -> `# fmt: on`
`# ruff:noqa` --> `# ruff: noqa`
`# ruff:noqa:F811` --> `# ruff: noqa: F811` (or at least `# ruff: noqa:F811` if generalized)
`# flake8:noqa:F811` --> `# flake8: noqa: F811` (or at least `# flake8: noqa:F811` if generalized)
`# ruff:isort:skip_file` -> `# ruff: isort: skip_file` (or at least `# ruff: isort:skip_file` if generalized)
`# isort:skip_file` -> `# isort: skip_file`
...

Basically the same as https://docs.astral.sh/ruff/rules/missing-space-after-todo-colon/ but with a lot more keywords

Formatters (black and Ruff) already take of adding a space after `#`

**Comment by MichaReiser** [comment]
@JelleZijlstra what do you think of adding this to Black/Ruff's formatter?

**Comment by JelleZijlstra** [comment]
Yes, I was thinking about that. I feel it goes against the spirit of Black's AST safety check if we modify comments (it doesn't actually violate the check, of course, because comments aren't part of the AST). That concern becomes bigger the bigger the set of comments is for which we apply the transformation.

I think I'd be OK with normalizing whitespace in `# type:` comments in Black, because those are a standardized part of the language. I'd be hesitant to change `# fmt:off` comments because the point of those is to make Black *not* change things. With other tool-specific comments, I'd be a bit hesitant because we'd need to encode specific knowledge of what each tool accepts.

**Comment by AlexWaygood** [comment]
> With other tool-specific comments, I'd be a bit hesitant because we'd need to encode specific knowledge of what each tool accepts.

We'd need to do that for a ruff lint rule as well though, right?

**Comment by MichaReiser** [comment]
> Yes, I was thinking about that. I feel it goes against the spirit of Black's AST safety check if we modify comments (it doesn't actually violate the check, of course, because comments aren't part of the AST). That concern becomes bigger the bigger the set of comments is for which we apply the transformation.
I'm personally less concerned about that, but I know it is an important property of/for Black. What you mean by that is that it doesn't change the representation as Python understands the AST? I'm saying that because our AST has much more details than Python's, e.g. we represent implicit concatenated strings and Black's unstable string-processing changes our AST (but presumably not Python's). One exception that Ruff already does here today is that it formats code blocks in docstrings.

> I'd be hesitant to change # fmt:off comments because the point of those is to make Black not change things.

That makes sense. Although I could see us formatting it as well, considering that [they get correctly indented today](https://black.vercel.app/?version=stable&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ACiAGNdAD2IimZxl1N_WlbvK5V_5PM4Nceez9J8N3C1qIg8e7k4xLRet8g5rDCr1ZTcLnwPqqNluQY86f5RteS3yDZKrumMMSNP1D3BaaIGngwC75nTfhipsVg_PwJW6isSDNnkYGp7AAAAIFUG6pwdLrEAAX-jAQAAAOrpGraxxGf7AgAAAAAEWVo=). 

> I'd be a bit hesitant because we'd need to encode specific knowledge of what each tool accepts

That's fair. I'm not too familiar with any of those tools so it's hard for me to judge how common it is that the suppression comment format changed in the past. One argument I see against it is that Black so far has avoided very specific formatting rules to keep its formatting predictable (which I consider a key difference to Prettier, which has tons of very specific rules). Adding these tool-specific formatting rules goes against this.

> We'd need to do that for a ruff lint rule as well though, right?

The difference that I see here is that you can opt out of the linter in case your tool changes in an incompatible way but you can't really do this in the formatter.

**Comment by JelleZijlstra** [comment]
> We'd need to do that for a ruff lint rule as well though, right?

As @MichaReiser said, a lint rule is more granular while a formatter is pretty much all-or-nothing. (At least, Black is; I believe your formatter is a little more configurable.) With a lint rule you could add a configuration option to allow people to control what set of comments get formatted.

**Comment by Avasam** [comment]
> you could add a configuration option to allow people to control what set of comments get formatted.

Good point. Ruff could add defaults for all the special comments it supports (see my list above) and let users expand with their other tools (up to you if you want to support `# type:` and `# pyright:` by default, since these are outside Ruff, but popular)

**Comment by egberts** [comment]
Latest PyCharm fights with RUFF over their preferred format of RUF100 with this single space between RUF100 and colon symbol.

PyCharm only recognizes this format but RUF complains:
```Python
call_me()  # NOQA : RUF100
```

RUF reformats as :
```Python
call_me()  # NOQA: RUF100
```
but PyCharm complains:

Nobody is happy.

**Comment by dhruvmanila** [comment]
@egberts Can you say more about this? What do you mean by "PyCharm only recognizes this format (`NOQA : RUF100`)" ?

<img width="678" alt="Screenshot 2024-07-09 at 09 42 50" src="https://github.com/astral-sh/ruff/assets/67177269/26470b48-dc93-45ca-a652-508743b1c378">

**Comment by egberts** [comment]
It's a battle between RUFF and PyCharm,  one wants a space between the NOQA and the semicolon, the other doesn't.

After letting RUFF making the fix, PyCharm complains again.  Comply with PyCharm and RUFF  fix it up right back.

**Comment by charliermarsh** [comment]
I think the confusion on my end is that I've never seen PyCharm enforce this and I use it daily. I would be extremely surprised if PyCharm required a space before a colon here. Are you sure you don't have some specific configuration that's causing this?

**Comment by Avasam** [comment]
https://github.com/astral-sh/ruff/pull/14111#issuecomment-2553636322 from @MichaReiser 
> I'll close this PR because there's no agreement on the rule's semantics. We should define and align on that first before working on a new PR.
> 
> Thank you @InSyncWithFoo for initiating the discussion and creating this pr.

I'd like to bring back my above comment:

> > you could add a configuration option to allow people to control what set of comments get formatted.
> 
> Good point. Ruff could have defaults for all the special comments it supports (see my list above) and let users expand with their other tools (up to maintainers if they want to support other popular tool comments by default)
> 
> With regex (maybe glob?) support, this could technically superseed [docs.astral.sh/ruff/rules/missing-space-after-todo-colon](https://docs.astral.sh/ruff/rules/missing-space-after-todo-colon/) entirely.

Basically, have a way to identify which comments are considered special and *require* a space after `:`. Completely generalising [missing-space-after-todo-colon (TD007)](https://docs.astral.sh/ruff/rules/missing-space-after-todo-colon/#missing-space-after-todo-colon-td007), and avoiding forcing Ruff to go down the rabbit hole of supporting every special comment ever.

Following existing rule configurations, it could be a list of regexes, supporting an `extend-*` config.
If you don't want regexes, we could leave `TD007` as a separate rule that looks at `task-tags` for matching optional parenthesis (`# TODO (Ava): blablabla`)

There's good arguments above (as well as the fact I wanna make it configurable) as to why this shouldn't be handled by the formatter.

---

As for naming, which is always the hard part:

Code would be `RUF1XX` (pragma related rules)
Rule name could be:
- `missing-space-after-pragma-colon` (my favorite)
- `missing-space-after-directive-colon` ("directive" matches naming of `rst-directive-colons` in https://github.com/pre-commit/pygrep-hooks, which also has some pragme related rules that made it to Ruff)
- `unspaced-pragma-comment`
- `pragma-comment-spacing`

For config, I suggest:
- `lint.ruff.pragma-tags` (group-scoped)
- `lint.pragma-tags` (mirrors the already top-level `task-tags`)
